File: spec/functional_test/fixtures/java/spring/src/GatewayRouteConfig.java
package org.test;

import org.springframework.cloud.gateway.handler.predicate.PredicateSpec;
import org.springframework.cloud.gateway.route.RouteLocator;
import org.springframework.cloud.gateway.route.builder.RouteLocatorBuilder;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpMethod;

@Configuration
public class GatewayRouteConfig {

    private static final String MCP_ENDPOINT_PATH = "/gateway/mcp";

    @Bean
    public RouteLocator customRouteLocator(RouteLocatorBuilder builder) {
        return builder.routes()
            .route("post_mcp", r -> r.method(HttpMethod.POST).and().path(MCP_ENDPOINT_PATH).uri("no://op"))
            .route("get_report", r -> isGetReportRoute(r).uri("no://op"))
            .build();
    }

    private PredicateSpec isGetReportRoute(PredicateSpec predicateSpec) {
        return predicateSpec.method(HttpMethod.GET).and().path("/gateway/reports/{id}");
    }
}

File: spec/functional_test/fixtures/kotlin/spring/src/main/kotlin/com/example/blog/GatewayRouteConfig.kt
package com.example.blog

class GatewayRouteConfig {
    fun customRouteLocator(builder: RouteLocatorBuilder): RouteLocator {
        val routesBuilder = builder.routes()
        routesBuilder.route("post") { predicateSpec ->
            predicateSpec
                .order(0)
                .isPostRequestToMcpEndpoint().and()
                .uri("no://op")
        }
        routesBuilder.route("get") { predicateSpec ->
            predicateSpec.isGetRequestToMcpEndpoint().uri("no://op")
        }
        routesBuilder.route("delete") { predicateSpec ->
            predicateSpec.isDeleteRequestToMcpEndpoint().uri("no://op")
        }
        return routesBuilder.build()
    }

    private fun PredicateSpec.isPostRequestToMcpEndpoint() =
        method(HttpMethod.POST).and().path(GatewayPolicy.MCP_ENDPOINT_PATH)

    private fun PredicateSpec.isGetRequestToMcpEndpoint() =
        method(HttpMethod.GET).and().path(GatewayPolicy.MCP_ENDPOINT_PATH)

    private fun PredicateSpec.isDeleteRequestToMcpEndpoint() =
        method(HttpMethod.DELETE).and().path(GatewayPolicy.MCP_ENDPOINT_PATH)
}
